### PR TITLE
StudentT

### PR DESCRIPTION
**Pull request by kydreth**
StudentT Distribution

Adding to the sampling mean/standard deviation need for a common function call. I had to use 10 million samples for accurate mean / standard deviation

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	/**
+	 * Computer Generation of Statistical Distributions
+	 * by Richard Saucier
+	 * ARL-TR-2168 March 2000
+	 * 5.1.23 page 36
+	 */
+	public static double sample(double v, Random random) {
+		assert( v > 0. );
+		return Gaussian.sample( 0., 1., random) / sqrt( ChiSquared.sample((int) v, random) / v );
+
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Nit: White space

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+public class StudentT {
+	/**
+	 * Computer Generation of Statistical Distributions
+	 * by Richard Saucier
+	 * ARL-TR-2168 March 2000
+	 * 5.1.23 page 36
+	 */
+	public static double sample(double v, Random random) {
+		assert( v > 0. );
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: No need for assert. If it's necessary, throw an IllegalArgumentException with an appropriate message saying what the problem is.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		return Gaussian.sample( 0., 1., random) / sqrt( ChiSquared.sample((int) v, random) / v );
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Should v just be an int?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		
+	}
+	
+	public static double pdf(double v, double t) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Should v be an Int?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		double halfV = v / 2.;
+		double numerator = gamma(halfVPlusOne);
+		double denominator = sqrt(v * PI) * gamma(halfV);
+		double multiplier = pow(1. + (pow(t, 2) / v), -halfVPlusOne);
+		
+		return (numerator / denominator) * multiplier;
+	}
+	
+	public static double logPdf(double v, double t) {
+		return log(pdf(v, t));
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Is there not a more simple function for log of the pdf? Are there any problems, numerical stability-wise, when t is really big?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	
+	private static double zeroDerivativeOfDigamma(double in) {
+		return digamma(in);
+	}
+	
+	public static Diff dPdf(double v, double t) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Impressive work here man!

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		double gammaHalfV = gamma(v / 2.);
+		double gammaHalfVPlusOne = gamma((v + 1.) / 2.);
+		double sqrtVPi = sqrt(v * PI);
+		double tSq = pow(t, 2);
+		double tSqDividedByVPlusOne = (tSq / v) + 1.;
+		double sqrtVPiGammaHalfV = sqrtVPi * gammaHalfV;
+		double sqrtPiGammaHalfV = sqrt(PI) * gammaHalfV;
+		
+		double zeroDerivativeOfDigammaHalfV = zeroDerivativeOfDigamma(v / 2.);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: The zeroDerivativeOfDigamma is just digamma right? Would it be clearer to just use digamma?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+		double dPdv = dPdv_multiplier * (dPdv_pt1 + dPdv_pt2 - dPdv_pt3 + dPdv_pt4);
+		
+		double dPdt_numerator = (t * (-v - 1)) * pow(tSqDividedByVPlusOne, -v / 2.) * gammaHalfVPlusOne;
+		double dPdt_denominator = (tSq + v) * sqrt((tSq + v) / v) * sqrtVPiGammaHalfV;
+		double dPdt = dPdt_numerator / dPdt_denominator;
+		
+		return new Diff(dPdv, dPdt);
+	}
+	
+	public static Diff dLogPdf(double v, double t) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Nicely done here too!

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+public class StudentTVertex extends ProbabilisticDouble {
+	
+	private final DoubleVertex v;
+	private final Random random;
+	
+	public StudentTVertex(DoubleVertex v, Random random) {
+		this.v = v;
+		this.random = random;
+		setValue(sample());
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java]: No need to set an initial value from sample.

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	@Override
+	public Map<String, Double> dLogPdf(Double value) {
+		StudentT.Diff diff = StudentT.dLogPdf(v.getValue(), value);
+		return convertDualNumbersToDiff(diff.dPdv, diff.dPdt);
+	}
+	
+	private Map<String, Double> convertDualNumbersToDiff(double dPdv, double dPdt) {
+		PartialDerivatives dPdInputsFromV = v.getDualNumber().getPartialDerivatives().multiplyBy(dPdv);
+		PartialDerivatives dPdInputs = dPdInputsFromV;
+		dPdInputs.asMap().put(getId(), dPdt);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java]: dPdInputs.putWithRespectTo(getId(), dPdt)

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.PartialDerivatives;
+
+import java.util.Map;
+import java.util.Random;
+
+public class StudentTVertex extends ProbabilisticDouble {
+	
+	private final DoubleVertex v;
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java]: I think this is an IntegerVertex if v is always an integer.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+	private double sigma = 1.;
+	
+	@Before
+	public void setup() {
+		random = new Random(1);
+	}
+	
+	@Test
+	public void sampleTest() {
+		int N = 10000000;
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java]: That's surprising you need 10million samples to get within 0.1. Is it both the std and mean?

Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java]: Also, does the TDistribution from apache math need this many samples?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java (hunk 6 above)
Comment (kydreth) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Thanks to @GeorgeNash

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java (hunk 7 above)
Comment (kydreth) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: All the maths refer to zero derivative of digamma, so I wasn't sure if that would remove clarity

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java (hunk 12 above)
Comment (kydreth) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java]: I don't think the apache one doesn't, ours did. I tried 100k, 1 mil, 10 mil. There was a huge variance with the std deviation.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+	public static Diff dPdf(int v, double t) {
+		double gammaHalfV = gamma(v / 2.);
+		double gammaHalfVPlusOne = gamma((v + 1.) / 2.);
+		double sqrtVPi = sqrt(v * PI);
+		double tSq = pow(t, 2);
+		double tSqDividedByVPlusOne = (tSq / v) + 1.;
+		double sqrtVPiGammaHalfV = sqrtVPi * gammaHalfV;
+//		double sqrtPiGammaHalfV = sqrt(PI) * gammaHalfV;
+//
+//		double zeroDerivativeOfDigammaHalfV = zeroDerivativeOfDigamma(v / 2.);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: Can we remove this commented out code?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+	 *
+	 * @param v Degrees of Freedom
+	 * @param t random variable
+	 * @return Differential of the Log of the Probability Density Function
+	 */
+	public static Diff dLogPdf(int v, double t) {
+		double tSq = pow(t, 2);
+		double tSqPlusV = tSq + v;
+		
+//		double digammaHalfV = zeroDerivativeOfDigamma(v / 2.);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/continuous/StudentT.java]: And this commented out code?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+		Map<String, DoubleTensor> m = new HashMap<>();
+		m.put(getId(), DoubleTensor.scalar(diff.dPdt));
+		return m;
+	}
+	/**
+	 *
+	 * @param dPdt d/dt
+	 * @return Map<String, Double> of the derivatives
+	 */
+	private Map<String, DoubleTensor> convertDualNumbersToDiff(double dPdt) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertex.java]: You don't need this method anymore right? Can we remove it?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+import static java.lang.Math.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+import static junit.framework.TestCase.assertEquals;
+
+/**
+ *
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java]: Remove or add comments

Comment (kydreth) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/probabilistic/StudentTVertexTest.java]: I believe this is done, please advise if not.

**Pull request merged**